text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,191 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or runtime), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.
+
+This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.
+
+WASM, asm.js and embedded platforms are excempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
+
+A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is subject to bikeshedding.
+
+The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.
+
+The JSON schema specifying the format is provided below. If you find Rust structures more readable, you can find them [here](https://github.com/rust-secure-code/cargo-auditable/blob/311f9932128667b8b18113becdea276b3d98aace/auditable-serde/src/lib.rs#L99-L172). In case of divergences the JSON schema provided in this RFC takes precedence.
+
+```json
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
+    "required": [
+        "packages"
+    ],
+    "properties": {
+        "packages": {
+            "type": "array",
+            "items": {
+                "$ref": "#/definitions/Package"
+            }
+        }
+    },
+    "definitions": {
+        "DependencyKind": {
+            "type": "string",
+            "enum": [
+                "build",
+                "runtime"
+            ]
+        },
+        "Package": {
+            "description": "A single package in the dependency tree",
+            "type": "object",
+            "required": [
+                "name",
+                "source",
+                "version"
+            ],
+            "properties": {
+                "dependencies": {
```

Comment:
Another piece of information that may be useful is the flags provided to the compiler (which would allow checking if various mitigations, like control-flow guard and retpoline, were enabled).

Reply:
This can already be detected by dedicated tools without requiring this information to be explicitly recorded at compile time: https://github.com/slimm609/checksec.sh https://github.com/trailofbits/winchecksec etc.